lib/explorer/shared.ex

```diff
@@ -270,71 +274,176 @@ defmodule Explorer.Shared do
 
   If no preferred type is given (nil), then the inferred type is returned.
   """
-  def dtype_from_list!(_list, :null), do: :null
+  def dtype_from_list!(list, preferred_type, strict \\ false)
+  def dtype_from_list!(_list, :null, _strict), do: :null
 
-  def dtype_from_list!(list, nil), do: dtype_from_list!(list)
+  # , do: dtype_from_list!(list, strict)
+  def dtype_from_list!(list, nil, strict) do
+    list
+    |> Enum.reduce(:null, &infer_type(&1, &2, nil, strict))
+    |> normalise_dtype!()
+  end
 
-  def dtype_from_list!(list, preferred_type) do
+  def dtype_from_list!(list, preferred_type, strict) do
     list
-    |> dtype_from_list!()
+    |> Enum.reduce(:null, &infer_type(&1, &2, preferred_type, strict))
     |> merge_preferred(preferred_type)
   end
 
   @non_finite [:nan, :infinity, :neg_infinity]
 
-  defp infer_type(nil, type), do: type
-  defp infer_type(item, :null), do: infer_type(item)
-  defp infer_type(integer, {:f, 64}) when is_integer(integer), do: {:f, 64}
-  defp infer_type(float, {:s, 64}) when is_float(float) or float in @non_finite, do: {:f, 64}
-  defp infer_type(list, {:list, type}) when is_list(list), do: infer_list(list, type)
-  defp infer_type(%{} = map, {:struct, inner}), do: infer_struct(map, inner)
+  defp infer_type(nil, type, _preferred, _strict), do: type
+  defp infer_type(item, :null, preferred, strict), do: infer_type(item, preferred, strict)
+
+  defp infer_type(integer, {:u, 64}, _preferred, _strict)
+       when is_integer(integer) and integer > 0,
+       do: {:u, 64}
```

Comment:
Do we need this clause? I think we can remove it and let it fallback to infer_type/3.